src/lib/shady-render.ts

```diff
@@ -22,31 +23,83 @@ declare global {
+  const cacheKey = `${result.type}--${scopeName}`;
+  let templateCache = templateCaches.get(cacheKey);
+  if (templateCache === undefined) {
+    templateCache = new Map<TemplateStringsArray, Template>();
+    templateCaches.set(cacheKey, templateCache);
+  }
+  let template = templateCache.get(result.strings);
+  if (template === undefined) {
+    const element = result.getTemplateElement();
+    window.ShadyCSS.prepareTemplateDom(element, scopeName);
+    template = new Template(result, element);
+    templateCache.set(result.strings, template);
+  }
+  return template;
+};
 
-        if (typeof window.ShadyCSS === 'object') {
-          window.ShadyCSS.prepareTemplate(element, scopeName);
-        }
+const TEMPLATE_TYPES = ['html', 'svg'];
+function removeStylesFromLitTemplates(scopeName: string) {
+  TEMPLATE_TYPES.forEach((type) => {
+    const templates = templateCaches.get(`${type}--${scopeName}`);
+    if (templates) {
+      templates.forEach((template) => {
+        const {element: {content}} = template;
+        const styles = content.querySelectorAll('style');
+        removeNodesFromTemplate(template, Array.from(styles));
+      });
+    }
+  });
+}
 
-        template = new Template(result, element);
-        templateCache.set(result.strings, template);
-      }
-      return template;
-    };
+const shadyRenderSet = new Set<string>();
+
+function hostForNode(node: Node) {
+  return node.nodeType === Node.DOCUMENT_FRAGMENT_NODE && (node as ShadowRoot).host
+}
 
 export function render(
     result: TemplateResult,
     container: Element|DocumentFragment,
     scopeName: string) {
-  return baseRender(result, container, shadyTemplateFactory(scopeName));
+  const host = hostForNode(container);
+  if (host && typeof window.ShadyCSS === 'object') {
+    const templateFactory = shadyTemplateFactory(scopeName);
+    const renderer = (container: Element|DocumentFragment, fragment: DocumentFragment) => {
+      if (!shadyRenderSet.has(scopeName)) {
+        shadyRenderSet.add(scopeName);
+        const styleTemplate = document.createElement('template');
+        Array.from(fragment.querySelectorAll('style')).forEach((s: Element) => {
+          styleTemplate.content.appendChild(s);
+        });
+        window.ShadyCSS.prepareTemplateStyles(styleTemplate, scopeName);
+        // fix templates.
+        removeStylesFromLitTemplates(scopeName);
+        // ApplyShim case.
+        if (window.ShadyCSS.nativeShadow) {
+          const style = styleTemplate.content.querySelector('style');
+          if (style) {
+            // insert style into rendered fragment
+            fragment.insertBefore(style, fragment.firstChild);
+            // insert into lit-template (for subsequent renders)
+            const template = templateFactory(result);
+            insertNodeIntoTemplate(template, style.cloneNode(true),
+                template.element.content.firstChild);
+          }
+        }
+      }
+      window.ShadyCSS.styleElement(host);
```

Comment:
For the first time calling `ShadyCSS.styleElement()` is ok, but if a change is made to the template and value of CSS custom property is changed, `ShadyCSS.styleSubtree()` has to be called to update nested elements using those CSS custom properties.

For example: 
```javascript
template = (active) => html`
  <style>
     my-button {
       --custom-color: red;
     }
     my-button.active {
       --custom-color: blue;
     }
  </style>
  <my-button class="${active ? 'active' : ''}"></my-button>
`;
```
Without calling `styleSubtree`, after change `active` to `true` in IE11 `<my-button>` will be still red.

Reply:
There's not a spot to do this in lit-html, but it does deserve documentation since it is necessary. It will be handled separately in LitElement.

Reply:
But there is a spot... You actually put `window.ShadyCSS.styleElement(host)` here. It has to be simply changed to `styleSubtree` to update all nested elements as well. I think it is very low-level API and LitElement using lit-html won't be able to do anything with that.

Reply:
The call to `styleElement` is only made when a lit `TemplateInstance` is created, not updated. Typically this is first render only, and this is done for efficiency. Therefore, changing this to `styleSubtree` wouldn't help with the case outlined here.

In general, ShadyCSS [requires](https://github.com/webcomponents/shadycss#usage) calling `ShadyCSS.styleSubtree` when dynamic updates that affect custom properties are made. This is specifically for the custom property shim which is only needed on IE11.

Calling `ShadyCSS.styleSubtree` is not cheap and doing so on every call to `shady-render` would have a pretty negative performance impact. For this reason, it seems better **not** to handle this case generally in `shady-render` and instead require users to call `ShadyCSS.styleSubtree` when needed. This is not ideal but seems like the best tradeoff.